### ISSUE TITLE
Namerd watch stopped updating one service on one namerd.

### ISSUE THREAD

**Issue by taer** [linkerd/linkerd#2121]
Issue Type:
- Bug report

**What happened**:
One of our 6 namerd servers suddenly stopped updating it's consul watch for just one service

**What you expected to happen**:
The namerd should continue to update it's client state from consul

**How to reproduce it (as minimally and precisely as possible)**:
Seems to be somewhat related to large app deploys. The app in question that had issues has this deployment strategy
* 120 apps running with a tag of `live` in consul
* namerd returns all apps with `live` tag
* 120 additional apps deployd(240 total now). 120 `live`, 120 not-live
* new instances are tagged `live`  (240 live now)
* once new apps are taking traffic and look clean, the old apps are untagged live (120 live, 120 non-live)
* the old apps are removed(from consul and existence).

**Anything else we need to know?**:
* I've attached a sanitized /namer_state/consul.json from the bad host. Called [totalWatch.json.gz](https://github.com/linkerd/linkerd/files/2367616/totalWatch.json.gz)
* I've attached a sanitized /namer_state/consul.json from a good host. Called [goodNamerd.json.gz](https://github.com/linkerd/linkerd/files/2367660/goodNamerd.json.gz)
* The bad service is named `/.local/live/serviceName-production`
* There were a few other tagged deployments of the bad app. If you search for `serviceName-production`, you'll see all the other versions
* There is no reference to the appName in the namerd logs.
* The namerd logs do contain `com.twitter.finagle.ChannelClosedException: ChannelException at remote address: /127.0.0.1:8500. Remote Info: Not Available`, but this appears at the same frequency on the good namerd vs the old one. Consul is running on port `8500` on localhost
* It is `running: true`, but it's last updateAt was `2018-09-05 21:08:52 +0000`
* This was captured around `2018-09-07 08:16:23.098530629 -0500`
* The log file for the bad namerd is also attached. Only sanitation was `D 0905 21:05:16.374 UTC THREAD32: consul lookup: .local /#/consul/.local/ver-0.50.176-da1c896420/removeAppName-production`

On the other 5 namerds, the lastUpdated times were:
```
      "lastUpdatedAt": "2018-09-07 13:16:25 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:25 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:25 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:51 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:51 +0000",
```

**Environment**:
linkerd and namerd 1.4.5

[totalWatch.json.gz](https://github.com/linkerd/linkerd/files/2367616/totalWatch.json.gz)
[goodNamerd.json.gz](https://github.com/linkerd/linkerd/files/2367660/goodNamerd.json.gz)
[badNamerd.log.cleaned.gz](https://github.com/linkerd/linkerd/files/2367705/badNamerd.log.cleaned.gz)

**Comment by adleong** [comment]
Thanks for all the great info in this issue, @taer!

**Comment by peterfroehlich** [comment]
We are running in the same problems, but with serverset namers. Sometimes a namerd node stays responsive but does not recieve updates from zookeeper anymore. Logs are clean.

**Comment by dadjeibaah** [comment]
@peterfroehlich thanks for letting us no, are you able to provide watch states for your namerd instances that are running into this issue?

**Comment by peterfroehlich** [comment]
Not yet, but I will build something that writes everything down when the error happens again.

**Comment by peterfroehlich** [comment]
We ran into this problem again, and this time I looked through all documented admin endpoints. I found a Deadlock in the Contention output on the broken node: 

```
DEADLOCKS:

"finagle/netty4-1-EventThread" Id=52 BLOCKED on com.twitter.finagle.serverset2.Zk2Resolver@1fc72d68 owned by "finagle/netty4-1-EventThread" Id=28
	at com.twitter.finagle.serverset2.Zk2Resolver.$anonfun$addrOf_$9(Zk2Resolver.scala:225)
	-  blocked on com.twitter.finagle.serverset2.Zk2Resolver@1fc72d68
	at com.twitter.finagle.serverset2.Zk2Resolver.$anonfun$addrOf_$9$adapted(Zk2Resolver.scala:213)
	at com.twitter.finagle.serverset2.Zk2Resolver$$Lambda$834/781748775.apply(Unknown Source)
	at com.twitter.util.Witness$$anon$17.notify(Event.scala:454)
	at com.twitter.util.Event$$anon$8.$anonfun$register$13(Event.scala:220)
	at com.twitter.util.Event$$anon$8.$anonfun$register$13$adapted(Event.scala:211)
	at com.twitter.util.Event$$anon$8$$Lambda$847/1705443860.apply(Unknown Source)
	at com.twitter.util.Function$.$anonfun$synchronizeWith$1(Function.scala:41)
	-  locked java.lang.Object@5b5c78e7
	...



"finagle/netty4-1-EventThread" Id=28 BLOCKED on com.twitter.util.Var$$anon$4@7bbdbc52 owned by "finagle/netty4-7" Id=45
	at com.twitter.util.Var$$anon$4.observe(Var.scala:409)
	-  blocked on com.twitter.util.Var$$anon$4@7bbdbc52
	at com.twitter.util.Var$$anon$2.observe(Var.scala:68)
	at com.twitter.util.Var$$anon$2.observe(Var.scala:68)
	at com.twitter.util.Var$$anon$2.observe(Var.scala:68)
	at com.twitter.util.Var$$anon$2.observe(Var.scala:68)
	at com.twitter.util.Var$$anon$2.observe(Var.scala:68)
	at com.twitter.util.Var$$anon$2.$anonfun$observe$1(Var.scala:80)
	at com.twitter.util.Var$$anon$2.$anonfun$observe$1$adapted(Var.scala:68)
	...



"finagle/netty4-7" Id=45 BLOCKED on com.twitter.finagle.serverset2.Zk2Resolver@1fc72d68 owned by "finagle/netty4-1-EventThread" Id=28
	at com.twitter.finagle.serverset2.Zk2Resolver.$anonfun$addrOf_$9(Zk2Resolver.scala:225)
	-  blocked on com.twitter.finagle.serverset2.Zk2Resolver@1fc72d68
	at com.twitter.finagle.serverset2.Zk2Resolver.$anonfun$addrOf_$9$adapted(Zk2Resolver.scala:213)
	at com.twitter.finagle.serverset2.Zk2Resolver$$Lambda$834/781748775.apply(Unknown Source)
	at com.twitter.util.Witness$$anon$17.notify(Event.scala:454)
	at com.twitter.util.Event$$anon$8.$anonfun$register$13(Event.scala:217)
	at com.twitter.util.Event$$anon$8.$anonfun$register$13$adapted(Event.scala:211)
	at com.twitter.util.Event$$anon$8$$Lambda$847/1705443860.apply(Unknown Source)
	at com.twitter.util.Function$.$anonfun$synchronizeWith$1(Function.scala:41)
	-  locked java.lang.Object@6dbffb33
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@2e05833c
```

[contention-mm6.txt](https://github.com/linkerd/linkerd/files/2404416/contention-mm6.txt)
[contention-mm6.pprof.zip](https://github.com/linkerd/linkerd/files/2404419/contention-mm6.pprof.zip)
[stacks-mm6.txt](https://github.com/linkerd/linkerd/files/2404434/stacks-mm6.txt)

**Comment by dadjeibaah** [comment]
@peterfich this is great! Thanks for giving us the info. We will look into this.

**Comment by peterfich** [comment]
Hi @dadjeibaah,

I Think you have the wrong Peter.

**Comment by adleong** [comment]
@peterfroehlich thanks for reporting this but I believe it is a different issue from what @taer is reporting.  @taer is using Consul whereas the deadlock you describe is specific to Zookeeper.  I have opened https://github.com/linkerd/linkerd/issues/2129 to track the Zookeeper deadlock issue.

**Comment by adleong** [comment]
Hi @taer, sorry that it's taken me so long to dig in to this; thanks for your patience.

I'm looking at the namer state files you sent but I'm a bit confused by their contents.  `totalWatch.json`, which is the state from the bad namerd includes an entry for the service `/.local/live/serviceName-production` but `goodNamerd.json` does not.  Instead it only contains entries for services like `app39` etc.

Was the good namerd resolving requests for `/.local/live/serviceName-production`?  How do you determine which namerds are good vs bad?

**Comment by taer** [comment]
@adleong that was just an artifact on how I scrubbed the files for our security team.

The service on the bad namerd `local/live/serviceName-production` is present in the goodNamerds, but got randomly named something app-numbery

For the same service on the good ones, the lastUpdate times were
```      
"lastUpdatedAt": "2018-09-07 13:16:25 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:25 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:25 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:51 +0000",
      "lastUpdatedAt": "2018-09-07 13:16:51 +0000",
```

the same one in the badNamerd state had this
```
 "/.local/live/serviceName-production": {
    "state": {
      "running": true,
      "lastStartedAt": "2018-08-25 12:05:01 +0000",
      "lastStoppedAt": "2018-08-25 06:45:39 +0000",
      "lastUpdatedAt": "2018-09-05 21:08:52 +0000",
```

**Comment by adleong** [comment]
@taer ah, thanks for clarifying.

It seems like Namerd simply never gets a response from Consul, which causes it to get stuck like this.  I have not be able to reproduce this, however.  I have filed https://github.com/linkerd/linkerd/issues/2131 which would mitigate this by restarting watches after a certain timeout.

**Comment by adleong** [comment]
Fixed by #2131
